### PR TITLE
Make the network stack robust so that federations can survive network disruptions

### PR DESCRIPTION
**Pull request by elsirion**
Part of #15

### CODE REVIEW

**Comment by elsirion** [comment]
I think the second to last commit should run through in CI, just the last one that migrates our in-process integration tests to a mocked out networking stack seems to fail locally for me, maybe you got an idea @jkitman? If it doesn't work I'll split it off into a separate PR. I'll write some docs for the other commits today. Any feedback is very welcome :)

File: minimint/src/net/peers.rs [hunk 1]
```diff
@@ -0,0 +1,457 @@
...
+    ) {
+        let mut connection: Option<AnyFramedTransport<PeerMessage<M>>> = None;
+        let mut failed_reconnect_counter: u64 = 1;
+        let mut reconnect: Option<Instant> = Some(reconnect_time(failed_reconnect_counter));
+        let mut resend_queue = MessageQueue::default();
+        let mut last_received: Option<MessageId> = None;
+
+        let mut err = None;
+        loop {
+            if let Some(e) = err.take() {
```
Comment (elsirion) [inline minimint/src/net/peers.rs:249]: At some point I'd like to model that whole state machine in a nicer way, maybe with an enum? Idk yet.

File: minimint/src/net/peers.rs [hunk 2]
```diff
@@ -0,0 +1,457 @@
...
+
+    async fn timeout<F, T>(f: F) -> Option<T>
+    where
+        F: Future<Output = T>,
+    {
+        tokio::time::timeout(Duration::from_secs(100), f).await.ok()
+    }
+
+    #[tokio::test]
+    async fn test_connect() {
```
Comment (elsirion) [inline minimint/src/net/peers.rs:409]: a separate disconnect test would be cool eventually, but that needs a more versatile `MockNetwork` impl that supports killing connections.

File: minimint/src/net/peers.rs (hunk 1 above)
Comment (Maan2003) [reply minimint/src/net/peers.rs:249]: Yes, enums are very good for that. It also helps understand code better - explaining all possible states before the logic

**Review by Maan2003 — APPROVED** [review]
LGTM

File: minimint/src/net/connect.rs [hunk 3]
```diff
@@ -1,175 +1,255 @@
...
+    use std::fmt::Debug;
+    use std::future::Future;
+    use std::pin::Pin;
+    use std::sync::Arc;
+    use std::time::Duration;
+    use tokio::io::{DuplexStream, ReadHalf, WriteHalf};
+    use tokio::sync::mpsc::Sender;
+    use tokio::sync::Mutex;
+
+    pub struct MockNetwork {
```
Comment (jkitman) [inline minimint/src/net/connect.rs:109]: Is the advantage of using mocks that we don't need to depend on ports being open?

File: minimint/src/net/peers.rs [hunk 4]
```diff
@@ -0,0 +1,457 @@
...
+                        if let Err(e) = conn.send(PeerMessage {
+                            msg: umsg,
+                            ack: last_received
+                        }).await {
+                            err = Some(e);
+                            continue;
+                        };
+                    }
+                },
+                new_connection_res = incoming_connections.recv() => {
```
[the remark this answers is not in the document]
Comment (elsirion) [reply minimint/src/net/peers.rs:272]: My current plan is (pretty much as a next step after this PR) to implement a `AuthenticatedConnector` (that's why the current one is called `InsecureTcpConnector`). This means authentication would happen before we even construct our framed transport. I intend to use TLS, that's why I want to put it into the connector where we are still dealing with streams.

**Comment by MaxHillebrand** [comment]
Nice improvements!

I wonder, did you notice any issues in "test production"?

**Pull request merged**